### DISCUSSION TITLE
Thanks + A bunch of questions

### DISCUSSION THREAD

**Discussion by 0xeb** [OpenSalamander/salamander#47] [General]
Team,

Long time user here. Thanks for open sourcing.

- How active will you be to look at PRs?
- Are you okay if I start converting the code comments to English?
- Is Unicode support on the plans? I might want to tackle that first

**Comment by janrysavy** [comment]
We'll actively review PRs. You can start converting the code comments to English, some files are already translated or WIP, see pull requests.
Unicode is #1 request, it is on the roadmap for years. I did some experiments with W10+ UTF-8 support: https://forum.altap.cz/viewtopic.php?t=44639.

**Comment by 0xeb** [comment]
I saw that screenshot. Did you lose your work? is it checked-in somewhere in some branch?

I would love to tackle this task first.

**Comment by janrysavy** [comment]
I didn't publish this branch, I was hoping to translate the comments into English first to avoid conflicts.

**Reply by 0xeb** [reply]
Well, there seems to be PRs hanging in there since December and some for like a month.

How much deep into the Unicode changes you are?

Maybe push to a new branch called "Unicode", then we can take it from there.

I prefer to move fast and so on and the PR dance will slow me down.

I forked it, and I might just do changes on my own and take from your repo. Or just keep branching away and see where it goes.

**Comment by janrysavy** [comment]
There is also unresolved question regarding third-party libraries: https://github.com/OpenSalamander/salamander/issues/8

**Comment by melloware** [comment]
@0xeb welcome and I love your enthusiasm!  I have been a Salamander user since 1999 and love this tool so any contributions you make will be welcome!

**Comment by janrysavy** [comment]
There are two possible ways to implement Unicode support. The standard UTF-16 way is to convert strings from char to wchar_t and call the W versions of WinAPI. I chose experimentally the second way via UTF-8: https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page

I have no idea to what extent UTF-8 is supported in WinAPI, I don't know of any Salamander-like project that goes this path. So there is a real possibility that we will run into problems and revert to the standard wchar_t / UTF-16 solution instead of the UTF-8 path.

If you are interested in experimenting with the UTF-8, I will release my patches.

**Comment by 0xeb** [comment]
@janrysavy , I was actually thinking to go for the UTF-16 approach. Replace everything with TCHAR and support the `-DUNICODE` switch to allow builds for any target.

We will think how to deal with plugins as we go, even if we have to write a small compact layer to go from UTF-16 to UTF-8 if needed.

But I think cleaning the code and making it TCHAR compatible, is a good start.

I will try to tackle the comments translation this weekend, aim for everything at once. That will also get me read the code and get familiar with things.

**Comment by janrysavy** [comment]
UTF-16 is the preferred way, Windows uses UTF-16 internally. I'm not sure if it makes sense to go via TCHAR and support the ANSI version as well instead of going directly to wchar_t? Do we want or need to compile and test the ANSI version?

With the move to UTF-16, it would probably be worth considering paths handling code cleanup, encapsulate operations over the path in a dedicated object, and support long paths (MAX_PATH limit) as well, see https://github.com/OpenSalamander/salamander/issues/44

**Reply by 0xeb** [reply]
If we do the code correctly with TCHAR, we have no additional cost to compile like version 4.0 with no unicode support. We can always compile with `-DUNICODE` anyway and that means wchar_t everywhere and wide api calls.

And yes, UTF-16 is what Windows kernel uses internally, so we can go for that.

> With the move to UTF-16, it would probably be worth considering paths handling code cleanup, encapsulate operations over the path in a dedicated object, and support long paths (MAX_PATH limit) as well, see https://github.com/OpenSalamander/salamander/issues/44

Yes, that will be taken into consideration too.

**Reply by janrysavy** [reply]
I'm looking at [Windows Terminal](https://github.com/microsoft/terminal/tree/main) and it looks like it uses wchar_t, std::wstring, std::wstring_view, directly.

**Comment by janrysavy** [comment]
Regarding possible long paths support: recursive functions working with paths would need to use allocated buffers. This would be necessary because long paths would exceed the stack size limitations.

**Reply by 0xeb** [reply]
I will probably start using the STL when applicable. So an std::wstring will come in handy as needed. Same for automatic pointers (unique/shared_ptr, etc.). We will see.

**Reply by janrysavy** [reply]
Yes, STL, [WIL](https://github.com/microsoft/wil), [GSL](https://github.com/microsoft/GSL), maybe [{fmt}](https://github.com/fmtlib/fmt) instead of std::format, clang-format & clang-tidy.

**Comment by janrysavy** [comment]
We discussed the issue with @petrsol and we think that the problem should be divided into several simpler parts (pull requests):

1) conversion from ANSI to UTF-16 (char to wchar_t / std::wstring / std::wstring_view)
2) encapsulating path handling into an object
3) support for long paths

This solution would be clearer and more manageable in terms of reviewing pull requests. It would also allow implementation of point 2 before point 1.
